core/server/src/main/java/alluxio/worker/netty/DataServerReadHandler.java

```diff
@@ -0,0 +1,423 @@
+   */
+  private void replySuccess(Channel channel) {
+    channel.writeAndFlush(RPCProtoMessage.createOkResponse(null))
+        .addListeners(ChannelFutureListener.CLOSE_ON_FAILURE, new ChannelFutureListener() {
+          @Override
+          public void operationComplete(ChannelFuture future) throws Exception {
+            reset();
+          }
+        });
+  }
+
+  /**
+   * Returns true if the block read request is valid.
+   *
+   * @param request the block read request
+   * @return true if the block read request is valid
+   */
+  private String validateReadRequest(Protocol.ReadRequest request) {
+    if (mRequest == null) {
+      return "";
+    }
+
+    if (request.getCancel()) {
+      if (request.getId() != mRequest.mId) {
+        return String.format("The IDs do not match: [actual: %d, expected: %d].", mRequest.mId,
+            request.getId());
+      }
+      return "";
+    }
+    return String
+        .format("Received a read request %s on a busy channel (used by block %d).", request,
+            mRequest.mId);
+  }
+
+  /**
+   * Resets the handler to its initial state.
+   */
+  private void reset() {
+    if (mRequest != null) {
+      mRequest.close();
+      mRequest = null;
+    }
+    mLock.lock();
+    try {
+      mPacketReaderActive = false;
+      mPosToQueue = -1;
+      mPosToWrite = -1;
+    } finally {
+      mLock.unlock();
+    }
+  }
+
+  /**
+   * Checks whether this object should be processed by this handler.
+   *
+   * @param object the object
+   * @return true if this object should be processed
+   */
+  protected boolean acceptMessage(Object object) {
+    if (!(object instanceof RPCProtoMessage)) {
+      return false;
+    }
+    RPCProtoMessage message = (RPCProtoMessage) object;
+    return message.getType() == RPCMessage.Type.RPC_READ_REQUEST;
+  }
+
+  /**
+   * Initializes the handler for the given block read request.
+   *
+   * @param request the block read request
+   * @throws Exception if it fails to initialize
+   */
+  protected abstract void initializeRequest(Protocol.ReadRequest request) throws Exception;
+
+  /**
+   * Returns the appropriate {@link DataBuffer} representing the data to send, depending on the
+   * configurable transfer type.
+   *
+   * @param channel the netty channel
+   * @param len The length, in bytes, of the data to read from the block
+   * @return a {@link DataBuffer} representing the data
+   * @throws IOException if an I/O error occurs when reading the data
+   */
+  protected abstract DataBuffer getDataBuffer(Channel channel, long offset, int len)
+      throws IOException;
+
+  /**
+   * The channel handler listener that runs after a packet write is flushed.
+   */
+  private final class WriteListener implements ChannelFutureListener {
+    private final long mPosToWriteUncommitted;
+
+    /**
+     * Creates an instance of the {@link WriteListener}.
+     *
+     * @param pos the pos
+     */
+    public WriteListener(long pos) {
```

Comment:
Similar as the other comment reply. It represents the position to commit (flush).